File: src/logic/shapes/fill.zig
const Point = @import("../types.zig").Point;
const std = @import("std");
const Path = @import("paths.zig").Path;

pub const GradientStop = extern struct {
    color: [4]f32,
    offset: f32, // 0..1
};

const SerializedLinearGradient = struct {
    start: Point,
    end: Point,
    stops: []GradientStop,
};

const SerializedRadialGradient = struct {
    radius_ratio: f32,
    center: Point,
    destination: Point,
    stops: []GradientStop,
};

pub const SerializedFill = union(enum) {
    linear: SerializedLinearGradient,
    radial: SerializedRadialGradient,
    solid: [4]f32,
};

pub const LinearGradient = struct {
    start: Point,
    end: Point,
    stops: std.ArrayList(GradientStop),
};

pub const RadialGradient = struct {
    radius_ratio: f32,
    center: Point,
    destination: Point,
    stops: std.ArrayList(GradientStop),
};

pub const Fill = union(enum) {
    linear: LinearGradient,
    radial: RadialGradient,
    solid: [4]f32,

    pub fn new(input: SerializedFill, allocator: std.mem.Allocator) !Fill {
        return switch (input) {
            .solid => |color| {
                return Fill{
                    .solid = color,
                };
            },
            .linear => |gradient| {
                var stops = std.ArrayList(GradientStop).init(allocator);
                try stops.appendSlice(gradient.stops);

                if (stops.items.len > 10) {
                    @panic("Max stops number is 10!");
                }

                return Fill{
                    .linear = .{
                        .start = gradient.start,
                        .end = gradient.end,
                        .stops = stops,
                    },
                };
            },
            .radial => |gradient| {
                var stops = std.ArrayList(GradientStop).init(allocator);
                try stops.appendSlice(gradient.stops);

                if (stops.items.len > 10) {
                    @panic("Max stops number is 10!");
                }

                return Fill{
                    .radial = .{
                        .center = gradient.center,
                        .destination = gradient.destination,
                        .radius_ratio = gradient.radius_ratio,
                        .stops = stops,
                    },
                };
            },
        };
    }

    pub fn serialize(self: Fill) SerializedFill {
        return switch (self) {
            .solid => |color| {
                return SerializedFill{
                    .solid = color,
                };
            },
            .linear => |gradient| {
                return SerializedFill{
                    .linear = .{
                        .start = gradient.start,
                        .end = gradient.end,
                        .stops = gradient.stops.items,
                    },
                };
            },
            .radial => |gradient| {
                return SerializedFill{
                    .radial = .{
                        .center = gradient.center,
                        .destination = gradient.destination,
                        .radius_ratio = gradient.radius_ratio,
                        .stops = gradient.stops.items,
                    },
                };
            },
        };
    }
};

File: src/fill.ts
import { getCustomProgram, getCustomProgramId } from 'customPrograms'
import { Fill, ZigFill } from 'types'
import assertUnreachable from 'utils/assertUnreachable'

export function toFill(fill: ZigFill): Fill {
  if ('solid' in fill && fill.solid) {
    return { solid: [...fill.solid] }
  }
  if ('linear' in fill && fill.linear) {
    return {
      linear: {
        start: {
          x: fill.linear.start.x,
          y: fill.linear.start.y,
        },
        end: {
          x: fill.linear.end.x,
          y: fill.linear.end.y,
        },
        stops: [...fill.linear.stops].map((stop) => ({
          offset: stop.offset,
          color: [...stop.color],
        })),
      },
    }
  }
  if ('radial' in fill && fill.radial) {
    return {
      radial: {
        center: {
          x: fill.radial.center.x,
          y: fill.radial.center.y,
        },
        destination: {
          x: fill.radial.destination.x,
          y: fill.radial.destination.y,
        },
        radius_ratio: fill.radial.radius_ratio,
        stops: [...fill.radial.stops].map((stop) => ({
          offset: stop.offset,
          color: [...stop.color],
        })),
      },
    }
  }

  if ('program_id' in fill && typeof fill.program_id === 'number') {
    return {
      program: {
        code: getCustomProgram(fill.program_id).code,
        id: fill.program_id,
      },
    }
  }

  assertUnreachable(fill)
}

// Zigar input and outptu type ar not compatible
// ZigFill is comaptible with output, but input needs to specify only one value on enum,
// otherwise the error occurs "Only one property of SerializedFill can be given a value"
// That's why this function overrides types to ZigFIll
export function toZigFill(fill: Fill): ZigFill {
  if ('solid' in fill) {
    return { solid: [...fill.solid] } as ZigFill
  }

  if ('linear' in fill) {
    return {
      linear: {
        start: {
          x: fill.linear.start.x,
          y: fill.linear.start.y,
        },
        end: {
          x: fill.linear.end.x,
          y: fill.linear.end.y,
        },
        stops: [...fill.linear.stops].map((stop) => ({
          offset: stop.offset,
          color: [...stop.color],
        })),
      },
    } as ZigFill
  }

  if ('radial' in fill) {
    return {
      radial: {
        center: {
          x: fill.radial.center.x,
          y: fill.radial.center.y,
        },
        destination: {
          x: fill.radial.destination.x,
          y: fill.radial.destination.y,
        },
        radius_ratio: fill.radial.radius_ratio,
        stops: [...fill.radial.stops].map((stop) => ({
          offset: stop.offset,
          color: [...stop.color],
        })),
      },
    } as ZigFill
  }

  if ('program' in fill) {
    return {
      program_id: getCustomProgramId(fill.program.id, fill.program.code),
    } as ZigFill
  }

  assertUnreachable(fill)
}
